### PR TITLE
[Merged by Bors] - Native support for Gnosis Beacon Chain network

### PR DESCRIPTION
**Pull request by k1rill-fedoseev**
## Proposed Changes

Add a new hardcoded spec for the Gnosis Beacon Chain.
Ideally, official Lighthouse executables will be able to connect to the gnosis beacon chain from now on, using `--network gnosis` CLI option.

### CODE REVIEW

File: common/eth2_network_config/built_in_network_configs/gnosis/config.yaml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+# ---------------------------------------------------------------
+# Some forks are disabled for now:
+#  - These may be re-assigned to another fork-version later
+#  - Temporarily set to max uint64 value: 2**64 - 1
+
+# Altair
+ALTAIR_FORK_VERSION: 0x01000064
+ALTAIR_FORK_EPOCH: 512
+# Merge
+MERGE_FORK_VERSION: 0x02000064
```
Comment (michaelsproul) [inline common/eth2_network_config/built_in_network_configs/gnosis/config.yaml]:
```suggestion
BELLATRIX_FORK_VERSION: 0x02000064
```

File: common/eth2_network_config/built_in_network_configs/gnosis/config.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+MERGE_FORK_EPOCH: 18446744073709551615
```
Comment (michaelsproul) [inline common/eth2_network_config/built_in_network_configs/gnosis/config.yaml]:
```suggestion
BELLATRIX_FORK_EPOCH: 18446744073709551615
```

File: consensus/types/presets/gnosis/phase0.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+# 2**0 * 10**9 (= 1,000,000,000) Gwei
+EFFECTIVE_BALANCE_INCREMENT: 1000000000
+
+
+# Time parameters
+# ---------------------------------------------------------------
+# 2**0 (= 1) slots 12 seconds
+MIN_ATTESTATION_INCLUSION_DELAY: 1
+# 2**5 (= 32) slots 6.4 minutes
+SLOTS_PER_EPOCH: 32
```
Comment (michaelsproul) [inline consensus/types/presets/gnosis/phase0.yaml]: This preset value doesn't match the `GnosisEthSpec`.

Do you mind adding a test similar to `mainnet_presets_consistent` in `consensus/types/src/preset.rs` that will test this preset?

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Just a few comments from the first round of review, I'll come back for a more thorough review later

File: common/eth2_network_config/built_in_network_configs/gnosis/config.yaml (hunk 1 above)
Comment (michaelsproul) [reply common/eth2_network_config/built_in_network_configs/gnosis/config.yaml]: We snuck this rename in recently in ef7351ddfecbb35cd23dde11835d58d4d4e689a8, you'll need to rebase to bring it in

File: consensus/types/presets/gnosis/phase0.yaml (hunk 3 above)
Comment (k1rill-fedoseev) [reply consensus/types/presets/gnosis/phase0.yaml]: Added a test, should be consistent now

File: lcli/src/main.rs [hunk 4]
```diff
@@ -665,6 +665,7 @@ fn main() {
         .and_then(|eth_spec_id| match eth_spec_id {
             EthSpecId::Minimal => run(EnvironmentBuilder::minimal(), &matches),
             EthSpecId::Mainnet => run(EnvironmentBuilder::mainnet(), &matches),
+            EthSpecId::Gnosis => run(EnvironmentBuilder::gnosis(), &matches),
```
Comment (michaelsproul) [inline lcli/src/main.rs:668]: We also need to add `"gnosis"` to the `.possible_values` on line 41

File: common/eth2_network_config/built_in_network_configs/gnosis/config.yaml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+# 2**2 (= 4)
+INACTIVITY_SCORE_BIAS: 4
+# 2**4 (= 16)
+INACTIVITY_SCORE_RECOVERY_RATE: 16
+# 2**4 * 10**9 (= 16,000,000,000) Gwei
+EJECTION_BALANCE: 16000000000
+# 2**2 (= 4)
+MIN_PER_EPOCH_CHURN_LIMIT: 4
+# 2**16 (= 65,536)
+CHURN_LIMIT_QUOTIENT: 65536
```
Comment (michaelsproul) [inline common/eth2_network_config/built_in_network_configs/gnosis/config.yaml]:
```suggestion
CHURN_LIMIT_QUOTIENT: 4096
```

As per https://github.com/gnosischain/lighthouse-launch/blob/18fbd5dad0dec201a4547669af4425eaaa46ab94/config/config.yaml#L15

**Review by michaelsproul — CHANGES_REQUESTED** [review]

**Review by michaelsproul — APPROVED** [review]
I've synced GBC with this PR and it worked correctly.

I'll see if anyone else wants to review before we merge

**Review by pawanjay176 — APPROVED** [review]
I was able to sync the chain as well.

**Comment by paulhauner** [comment]
Looking good to me :relaxed: I think we should block this on the `v2.1.1` release (it's urgent).

Did anyone notice a difference in compile times or binary size due to the new configs?

**Comment by michaelsproul** [comment]
> Did anyone notice a difference in compile times or binary size due to the new configs?

Good point. I just checked:

* Without `gnosis` spec: 111MB binary, 2m41s compile time
* With `gnosis` spec: 132MB binary, 3m48s compile time

So that's an 18% increase in binary size and a 41% increase in compile time (ouch!). I think we should add a top-level `gnosis` feature that we enable for releases/CI builds, but which will be disabled by default for users building from source. Similar to the `spec-minimal` feature here:

https://github.com/sigp/lighthouse/blob/5f628a71d4b2a7e7761b30e726338bba07617cd2/lighthouse/Cargo.toml#L17-L18

**Comment by michaelsproul** [comment]
I noticed that we won't enable the gnosis feature when building the x86_64 `unstable` Docker image, due to some quirks of our Docker builds. I'm cleaning that up as part of #2938, so we can merge this now and come back for the images later?

**Comment by michaelsproul** [comment]
Dammit we've hit the license CLA bug. Sorry @k1rill-fedoseev do you mind pushing a new commit to trigger the license check again?

bors r-

File: lighthouse/Cargo.toml [hunk 6]
```diff
@@ -16,6 +16,8 @@ modern = ["bls/supranational-force-adx"]
 milagro = ["bls/milagro"]
 # Support minimal spec (used for testing only).
 spec-minimal = []
+# Support gnosis spec and gnosis beacon chain.
```
Comment (michaelsproul) [inline lighthouse/Cargo.toml]:
```suggestion
# Support Gnosis spec and Gnosis Beacon Chain.
```

Dummy edit to trigger CI again

Comment (michaelsproul) [reply lighthouse/Cargo.toml]: Ah, I can't commit this. Will need to wait for you Kirill :pray:

Comment (k1rill-fedoseev) [reply lighthouse/Cargo.toml]: I've committed this, but not sure if it helped. Do you see any changes regarding CLA?

**Pull request closed without merging**